pkg/syncer/cnsoperator/controller/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -0,0 +1,403 @@
+	k8sclient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("ValidateStorageQuotaforVMSnapshot2: Failed to initialize K8S client error: %v", err)
+
+		return err
+	}
+	for _, vmVolume := range vm.Spec.Volumes {
+		pvc, err = k8sclient.CoreV1().PersistentVolumeClaims(vm.Namespace).Get(ctx, vmVolume.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("validateStorageQuotaforVMSnapshot: get pvc %s/%s failed: %+v", vm.Namespace, vmVolume.Name, err)
+			return err
+		}
+		if pvc.Spec.VolumeName != "" {
+			pv, err = k8sclient.CoreV1().PersistentVolumes().Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+			if err != nil {
+				log.Errorf("validateStorageQuotaforVMSnapshot: could not get the volume for pvc %s", pvc.Name)
+				return err
+			}
+			if pv.Spec.CSI != nil && pv.Spec.CSI.VolumeHandle != "" {
+				cnsVolId := cnstypes.CnsVolumeId{Id: pv.Spec.CSI.VolumeHandle}
+				cnsVolumeIds = append(cnsVolumeIds, cnsVolId)
+				syncVolumeSpecs = append(syncVolumeSpecs, cnstypes.CnsSyncVolumeSpec{VolumeId: cnsVolId, SyncMode: syncMode})
+			}
+		} else {
+			msg := fmt.Sprintf("validateStorageQuotaforVMSnapshot: could not find the PV associated with PVC %s/%s", vm.Namespace, vmVolume.Name)
+			log.Error(msg)
+			return errors.New(msg)
+		}
+	}
+	if len(syncVolumeSpecs) == 0 {
+		log.Info("no volumes found for virtual machine %+v/%+v", vm.Name, vm.Namespace)
+		return nil
+	}
+
+	// Trigger CNS VolumeSync API for identified volume-lds and Fetch Latest Aggregated snapshot size
+	log.Infof("Trigger CNS VolumeSync API, for no of volumes %d", len(syncVolumeSpecs))
+	syncVolumeFaultType, err := r.volumeManager.SyncVolume(ctx, syncVolumeSpecs)
+	if err != nil {
+		log.Errorf("validateStorageQuotaforVMSnapshot: failed to sync volumes opertion "+
+			"failed with cnsfault: %q error %+v", syncVolumeFaultType, err)
+		return err
+	}
+	// fetch updated cns volumes
+	queryFilter := cnstypes.CnsQueryFilter{
+		VolumeIds: cnsVolumeIds,
+	}
+	queryResult, err := r.volumeManager.QueryVolume(ctx, queryFilter)
+	if err != nil {
+		log.Errorf("validateStorageQuotaforVMSnapshot: failed to query volume from cns with err: %v.", err)
+		return err
+	}
+	if queryResult != nil && len(queryResult.Volumes) > 0 {
+		for _, cnsvolume := range queryResult.Volumes {
+			val, ok := cnsvolume.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails)
+			if ok {
+				//  Update CNSVolumeInfo with latest aggregated Size and Update SPU used value.
+				patch, err := common.GetCNSVolumeInfoPatch(ctx, val.AggregatedSnapshotCapacityInMb, cnsvolume.VolumeId.Id) // TODO: UDPATE to value returned
+				if err != nil {
+					log.Errorf("validateStorageQuotaforVMSnapshot: failed to create cnsvolumeinfo patch error %+v", err)
+					return err
+				}
+				patchBytes, err := json.Marshal(patch)
+				if err != nil {
+					return err
+				}
+				err = r.volumeInfoService.PatchVolumeInfo(ctx, cnsvolume.VolumeId.Id,
+					patchBytes, allowedRetriesToPatchCNSVolumeInfo)
```

Comment:
Is this patch idempotent? let's suppose we patch cnsvolumeinfo, then the code after crashed, will patching it again be okay?

Reply:
yes patch is idempotent, as we handle the case by calculating the difference between old and new aggregated size and update the spu with only difference in values